### PR TITLE
[feature] exports rule-test-harness to be used with other frameworks and in-repo lint rules

### LINKED ISSUE #227: Expose infrastructure for testing custom rules.

### PR DESCRIPTION
**Pull request by gabrielcsapo**
fixes #227 

The usage for clients using mocha would be the following:

```js
'use strict';

const generateRuleTests = require('ember-template-lint/lib/rule-test-harness');

generateRuleTests({
  groupingMethod: describe,
  testMethod: it,
  name: 'no-unbound',

  config: true,

  good: ['{{foo}}', '{{button}}'],

  bad: [
    {
      template: '{{unbound foo}}',

      result: {
        message,
        moduleId: 'layout.hbs',
        source: '{{unbound foo}}',
        line: 1,
        column: 0,
      },
    },
    {
      template: '{{my-thing foo=(unbound foo)}}',

      result: {
        message,
        moduleId: 'layout.hbs',
        source: '(unbound foo)',
        line: 1,
        column: 15,
      },
    },
  ],
});
```

for clients using qunit they would be able to alter it slightly and it would become:

```js
'use strict';

const QUnit = require('qunit');
const generateRuleTests = require('ember-template-lint/lib/rule-test-harness');

generateRuleTests({
  groupingMethod: QUnit.module,
  testMethod: QUnit.test,
  name: 'no-unbound',

  config: true,

  good: ['{{foo}}', '{{button}}'],

  bad: [
    {
      template: '{{unbound foo}}',

      result: {
        message,
        moduleId: 'layout.hbs',
        source: '{{unbound foo}}',
        line: 1,
        column: 0,
      },
    },
    {
      template: '{{my-thing foo=(unbound foo)}}',

      result: {
        message,
        moduleId: 'layout.hbs',
        source: '(unbound foo)',
        line: 1,
        column: 15,
      },
    },
  ],
});
```

### CODE REVIEW

File: test/helpers/rule-test-harness.js [hunk 1]
```diff
@@ -1,176 +1,13 @@
...
-            delete expectedResults[i].message;
-          }
-        }
-
-        expect(actual).to.deep.equal(expectedResults);
-      });
-    });
-  });
+  return generateRuleTests(
+    Object.assign(options, {
```
Comment (rwjblue) [inline test/helpers/rule-test-harness.js]: To avoid mutating the passed in `options` let's do:

```suggestion
    Object.assign({}, options, {
```

**Comment by bmish** [comment]
Can you also update the plugin documentation page to document this and provide examples?

https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/plugins.md

**Comment by rwjblue** [comment]
Ya, good idea @bmish!

@gabrielcsapo - Mind doing that in a follow up? It will help make it clear that this is intended to be public API.

**Comment by gabrielcsapo** [comment]
Of course, will open up another RB for docs and have that up by tonight or
tomorrow morning!

**Pull request merged**